subsys/net/ip/net_pkt.c

```diff
@@ -519,14 +520,47 @@ static struct net_pkt *net_pkt_get(struct k_mem_slab *slab,
 	pkt = net_pkt_get_reserve(slab, net_if_get_ll_reserve(iface, addr6),
 				  timeout);
 #endif
-	if (pkt) {
+	if (pkt && slab != &rx_pkts) {
+		sa_family_t family;
+		uint16_t iface_len, data_len = 0;
+		enum net_ip_protocol proto;
+
 		net_pkt_set_context(pkt, context);
 		net_pkt_set_iface(pkt, iface);
 
-		if (context) {
-			net_pkt_set_family(pkt,
-					   net_context_get_family(context));
+		iface_len = net_if_get_mtu(iface);
+
+		family = net_context_get_family(context);
+		net_pkt_set_family(pkt, family);
+
+		if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
+			data_len = max(iface_len, NET_IPV6_MTU);
 		}
+
+		if (IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
+			data_len = max(iface_len, NET_IPV4_MTU);
+		}
+
+		proto = net_context_get_ip_proto(context);
+
+		if (IS_ENABLED(CONFIG_NET_TCP) && proto == IPPROTO_TCP) {
+			data_len -= NET_TCPH_LEN;
+			data_len -= NET_TCP_MAX_OPT_SIZE;
+		}
+
+		if (IS_ENABLED(CONFIG_NET_UDP) && proto == IPPROTO_UDP) {
+			data_len -= NET_UDPH_LEN;
+
+#if defined(CONFIG_NET_RPL_INSERT_HBH_OPTION)
```

Comment:
How about considering net_pkt_set_ipv6_ext_len() in case ipv6 and move away RPL check here. RPL setting this value when insert is success "net_pkt_set_ipv6_ext_len(pkt, ext_len + NET_RPL_HOP_BY_HOP_LEN);".

Reply:
RPL has not yet set the ext header length when this function is called.